src/cryptonote_core/cryptonote_tx_utils.cpp

```diff
@@ -226,35 +226,55 @@ namespace cryptonote
     std::vector<tx_extra_field> tx_extra_fields;
     if (parse_tx_extra(tx.extra, tx_extra_fields))
     {
+      bool add_dummy_payment_id = true;
       tx_extra_nonce extra_nonce;
       if (find_tx_extra_field_by_type(tx_extra_fields, extra_nonce))
       {
-        crypto::hash8 payment_id = null_hash8;
-        if (get_encrypted_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id))
+        crypto::hash payment_id = null_hash;
+        crypto::hash8 payment_id8 = null_hash8;
+        if (get_encrypted_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id8))
         {
-          LOG_PRINT_L2("Encrypting payment id " << payment_id);
+          LOG_PRINT_L2("Encrypting payment id " << payment_id8);
           crypto::public_key view_key_pub = get_destination_view_key_pub(destinations, change_addr);
           if (view_key_pub == null_pkey)
           {
             LOG_ERROR("Destinations have to have exactly one output to support encrypted payment ids");
             return false;
           }
 
-          if (!hwdev.encrypt_payment_id(payment_id, view_key_pub, tx_key))
+          if (!hwdev.encrypt_payment_id(payment_id8, view_key_pub, tx_key))
           {
             LOG_ERROR("Failed to encrypt payment id");
             return false;
           }
 
           std::string extra_nonce;
-          set_encrypted_payment_id_to_tx_extra_nonce(extra_nonce, payment_id);
+          set_encrypted_payment_id_to_tx_extra_nonce(extra_nonce, payment_id8);
           remove_field_from_tx_extra(tx.extra, typeid(tx_extra_nonce));
           if (!add_extra_nonce_to_tx_extra(tx.extra, extra_nonce))
           {
             LOG_ERROR("Failed to add encrypted payment id to tx extra");
             return false;
           }
-          LOG_PRINT_L1("Encrypted payment ID: " << payment_id);
+          LOG_PRINT_L1("Encrypted payment ID: " << payment_id8);
+          add_dummy_payment_id = false;
+        }
+        else if (get_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id))
+        {
+          add_dummy_payment_id = false;
+        }
+      }
+
+      if (add_dummy_payment_id)
+      {
+        // if we have neither long not short payment id, add a dummy short one,
+        // this should end up being the vast majority of txes as time goes on
+        std::string extra_nonce;
+        set_encrypted_payment_id_to_tx_extra_nonce(extra_nonce, crypto::rand<crypto::hash8>());
```

Comment:
It's random now instead of a fixed known value because I realized it meant a recipient could deduce whether at least one of the other outputs in that tx had a real payment id or not (if you get the special value, it means none of the other outputs have a payment id).

Reply:
That concern applies only to multi-destination transfers (i.e. 3 or more outputs including the change) where the use of payment ID doesn't make much sense (IIRC it's even disallowed by the wallet implementation). If it's a regular transfer to a single destination, the concern doesn't seem to apply and it seems to make sense to use the null payment ID.

Reply:
True, but then you end up with the odd special case that everyone will get wrong because it usually works in the common case.

Reply:
Assuming that the special case and the common case refer to multi-destination transfer and single-destination transfer, respectively, I find it hard to understand your comment. What do you mean by "everyone will get wrong"?

Reply:
I still don't understand what you mean. My above suggested code is triggered only when the tx is a standard single-destination transfer (where `get_destination_view_key_pub` returns non-null). Multi-destination transfers will use the randomly generated extra value, and each recipient will see a differently decrypted random payment ID.

Reply:
My point is that both Bob and Carol receive differently randomized payment IDs, so there’s nothing leaked. The use of encrypted null hash is only applicable to sending to a singke destination.

Reply:
So you're saying use a different rule for single vs multi destination tranfers ? That seems like a good compromise.

Reply:
Yes, as in my above suggested code.

Reply:
On second thoughts, it doesn't change much at all. You still get the special case effect most will get wrong.

Reply:
Can you describe the scenario using Alice et al?

Reply:
I did above:

Alice sends monero to Bob and Carol, without a payment id. She cannot include a short payment id that will decode to 00..00 for both. Thus, at least one of them will see a non 0 payment id (if using that dummy payment id system).

In your proposed case, both will.

Reply:
> You still get the special case effect most will get wrong.

What do you mean by this? And what’s the problem with my proposal of treating single-destination transfers as a special case?

Reply:
> You still get the special case effect most will get wrong.

What do you mean by this? And what’s the problem with my proposal of treating single-destination transfers as a special case?

Reply:
Since most txes have two outputs (including change), most recipients will get a zero payment id. In the rare case where someone receives an output from a 3 (or more) output tx, there's a 1/2 chance (or more, or 100% chance with your suggestion) they'll get a non zero one. My point is that people will code/test their stuff for 0 only since that's all they ever get if they test with 2 out txes.

Reply:
Valid point, but the more I think, the more I find the idea of using encrypted payment IDs for multi-destination txs nonsensical. Construction of such a tx is disabled officially anyway:

https://github.com/monero-project/monero/blob/e344d93ce706482fc55bb036051dc08da18fabbb/src/cryptonote_core/cryptonote_tx_utils.cpp#L236-L241

so we shouldn't need to worry about users relying on encrypted payment IDs being used in multi-destination txs. As such, it seems reasonable to make the wallet receiving fund in a multi-destination tx ignore the encrypted payment ID field and set the associated payment ID to zero regardless.

Reply:
I think I like the idea of disallowing payment ids for > 2 outs. I would not like the wallet disregarding any that arrives though. That sounds like it'll bite us in the rear later.

Reply:
> I think I like the idea of disallowing payment ids for > 2 outs.

It *is* already disallowed AFAIK.

> I would not like the wallet disregarding any that arrives though. That sounds like it'll bite us in the rear later.

Fair argument. Since the construction of multi-destination txs with encrypted payment IDs is already disallowed, the right solution seems to be to NOT add dummy payment IDs to multi-destination txs.

Reply:
Yes, I agree with that :)